### ISSUE TITLE
Remove `Gemfile.lock` from repo

### ISSUE THREAD

**Issue by kugaevsky** [chaadow/activestorage-openstack#14]
I believe that `Gemfile.lock` **must** be added to `.gitignore` for many reasons.
Here is the article from `bundler` maintainer Yehuda Katz https://yehudakatz.com/2010/12/16/clarifying-the-roles-of-the-gemspec-and-gemfile/.

To clarify the problem in our case, I had failed my deployments with the locked `fog-openstack` `0.2.2` version which has broken OpenStack Keystone v2 compatibility. Without gem version lock I could just update `fog-openstack` and fix my application, cause they resolve this issue very quickly.

**Comment by kugaevsky** [comment]
Created PR with fixed OpenStack auth for my provider: fog/fog-openstack#428
Hope you'll bump `fog-openstack` version after merge ASAP. 8)

**Comment by chaadow** [comment]
Please check the README of both this repository and `fog-openstack` 
https://github.com/chaadow/activestorage-openstack#migrating-from-fog-openstack-01x-to-02x

**Comment by chaadow** [comment]
If you want to use version 0.1.x of fog-openstack, in your gemfile, use ` gem 'activestorage-openstack, '~0.3'` @kugaevsky

**Comment by kugaevsky** [comment]
@chaadow Sorry, I don't see how migration linked to including `Gemfile.lock` into the repo?

**Comment by kugaevsky** [comment]
> If you want to use version 0.1.x of fog-openstack, in your gemfile, use gem 'activestorage-openstack, '~0.3' @kugaevsky

But it doesn't help me to set version of `fog-openstack` that I need in my application

**Comment by chaadow** [comment]
Gemfile.lock is needed for : 
- Compatibility of dependencies.
- Specifying a version of rails

**Comment by chaadow** [comment]
@kugaevsky actually it does. all you have to do is update your configuration file. everything is explained both in my gem and in fog-openstack. There is no breaking changes introduced in this gem. You need to configure your connection to fog-openstack properly.
The README helps and guides you to configure `fog-openstack` (which is not its responsibilig youty) while pointing you to `fog-openstack`'s README

**Comment by kugaevsky** [comment]
> Gemfile.lock is needed for :

Yes, in applications, but not in gems. Did you read Yehuda Katz's article in the first message?

How many mature gems with `Gemfile.lock` included do you know?

**Comment by chaadow** [comment]
Here is one : https://github.com/rails/rails

**Comment by chaadow** [comment]
This gem integrates with rails. so Gemfile.lock is needed for rails version and to be compatible with multiple version of rails.
For general purpose gems, you are right, Gemfile.lock should not be checked in the git repo

**Comment by kugaevsky** [comment]
If you want to tighten dependency your can set the exact version in `gemspec` file, but that is bad practice too.

> Here is one : https://github.com/rails/rails

That is a bad example, cause rails is multiple gems bundled together.

**Comment by kugaevsky** [comment]
> This gem integrates with rails. so Gemfile.lock is needed for rails version and to be compatible with multiple version of rails.

The thing is there will be only one compatible version in the current architecture. The one which is locked in `Gemfile.lock`.

**Comment by chaadow** [comment]
It is a bad practice for general purpose gems, again.
For rails plugins, Gemfile.lock is necessary

**Comment by chaadow** [comment]
Rails 6.0 is going to be released soon, which will lead to new changes to activestorage. this gem will then be compatible with it.

Let's say There is a fix only for rails version 5.2. 
1- I will create a branch based on the latest stable release of version 5.2 (let's say v0.3)
2 - fix the changes
3- Then release the branch with tag v0.3.1

That's how you maintain compatibility between multiple version of rails.

Hope that is clearer

**Comment by chaadow** [comment]
Here is an example of another activestorage adapter implementation : 
https://github.com/huacnlee/activestorage-aliyun

And it uses Gemfile.lock. Hope you catch the difference between "normal" gems and "rails plugins/adapters"

==> This is a "rails plugin" basically. so you need to enforce a version of rails ( at the very least)

**Comment by kugaevsky** [comment]
> That's how you maintain compatibility between multiple version of rails.
> Hope that is clearer

It's clear enough, thank you. But I believe that a better way to manage dependency versions in gems is a gemspec file, not `Gemfile.lock`. And it doesn't matter we are speaking about standalone gems or "rails plugins".

For example this code
```ruby
  s.add_dependency "fog-openstack", '~> 0.2.2'
  s.add_dependency "rails", "5.2.1"
```
will lock rails version to `5.2.1` and untie your hands to use any `fog-openstack` version in `0.2.2 => version < 0.3` range.

But with `Gemfile.lock` you **must** use the locked version of all gems from dependencies.

**Comment by kugaevsky** [comment]
In other words, you tightly lock you gem to dependencies (not only `rails`, but `fog-openstack` too), and you should release bumped version of gem in case of any patch release in `fog-openstack`.

**Comment by chaadow** [comment]
@kugaevsky Alright that's a valid point. that makes sense.

**Comment by kugaevsky** [comment]
@chaadow Should I make a PR? 8)

**Comment by chaadow** [comment]
@kugaevsky I've actually opened a PR to introduce TravisCI (which can be merged since it's working) so i'll add the commit there no worries.

**Comment by kugaevsky** [comment]
@chaadow Perfect! Thank you.

**Comment by chaadow** [comment]
@kugaevsky done. released v0.4.1 along with travis and test coverage niceties. enjoy
